lib/signalfx.js

```diff
@@ -0,0 +1,334 @@
+      _this.post(dataToSend, url);
+    }, 2).push();
+  };
+
+  SignalFxClient.prototype.post = function (data, postUrl, contentType) {
+    var _this = this;
+
+    var headers = {};
+    headers[_this.HEADER_USER_AGENT_KEY] = [name, version].join('/');
+    //Adding custom user agents passed by client modules
+    if (_this.userAgents) {
+      headers[_this.HEADER_USER_AGENT_KEY] += ',' + _this.userAgents;
+    }
+    headers[_this.HEADER_API_TOKEN_KEY] = _this.apiToken;
+    headers[_this.HEADER_CONTENT_TYPE] = contentType || _this.getHeaderContentType();
+
+    // An object of options to indicate where to post to
+    var postOptions = {
+      url: postUrl,
+      timeout: this.timeout,
+      headers: headers,
+      body: data,
+      method: 'POST'
+    };
+
+    request(postOptions, function (error, response, body) {
+      _this.asyncRunning = false;
+
+      if (error || response.statusCode !== 200) {
+        winston.info('Filed to sent datapoint: ', response.statusCode, body);
+      }
+    });
+  };
+
+
+// ============================================================================================
+//                  SignalFx API client with Protobuf data sending
+// ============================================================================================
+  /**
+   * SignalFx API client data handler that uses Protocol Buffers.
+   * This class presents the interfaces that handle the serialization of data
+   * using Protocol Buffers
+   *
+   * @param apiToken
+   * @constructor
+   */
+  function ProtoBufSignalFx(apiToken) {
+    SignalFxClient.apply(this, arguments);
+  }
+
+  ProtoBufSignalFx.prototype = Object.create(SignalFxClient.prototype);
+  ProtoBufSignalFx.prototype.constructor = SignalFxClient;
+
+  ProtoBufSignalFx.prototype.getHeaderContentType = function () {
+    return PROTOBUF_HEADER_CONTENT_TYPE;
+  };
+
+  ProtoBufSignalFx.prototype.addToQueue = function (metricType, datapoint) {
+    var protobufDatapoint = new protocolBuffers.DataPoint();
+
+    // Set value by type
+    protobufDatapoint.value = {};
+    if (typeof datapoint.value === 'string') { // string value
+      protobufDatapoint.value.strValue = datapoint.value;
+    } else if (typeof datapoint.value === 'number' && datapoint.value % 1 !== 0) { // 'double' value
+      protobufDatapoint.value.doubleValue = +datapoint.value;
+    } else if (typeof datapoint.value === 'number' && datapoint.value % 1 === 0) { // int value
+      protobufDatapoint.value.intValue = Math.round(+datapoint.value);
+    } else {
+      throw new TypeError('Invalid Value ' + datapoint.value);
+    }
+
+    protobufDatapoint.metricType = protocolBuffers.MetricType[metricType.toUpperCase()];
+    protobufDatapoint.metric = datapoint['metric'];
+
+    if (datapoint['timestamp']) {
+      protobufDatapoint.timestamp = +datapoint['timestamp'];
+    }
+
+    // set datapoint dimensions
+    var dimensions = datapoint['dimensions'] || [];
+    for (var key in dimensions) {
```

Comment:
are you expecting dimensions to be an array or an object? Because for..in is for accessing objects but in the previous line you're setting it to an array as a default.